src/python/gudhi/representations/vector_methods.py

```diff
@@ -807,20 +807,28 @@ def fit(self, X, y=None, sample_weight=None):
             self
         """
         if not hasattr(self.quantiser, 'fit'):
-            raise TypeError("quantiser %s has no `fit` attribute." % (self.quantiser))
-
-        # In fitting we remove infinite death time points so that every center is finite
-        X = [dgm[~np.isinf(dgm).any(axis=1), :] for dgm in X]
+            raise TypeError(f"quantiser {self.quantiser} has no `fit` attribute.")
 
         if sample_weight is None:
             sample_weight = [self.get_weighting_method()(measure) for measure in X]
 
         measures_concat = np.concatenate(X)
         weights_concat = np.concatenate(sample_weight)
-
-        self.quantiser.fit(X=measures_concat, sample_weight=weights_concat)
-
+        # In fitting we remove infinite birth/death time points so that every center is finite
+        filtered_measures_concat = measures_concat[~np.isinf(measures_concat).any(axis=1), :] if len(measures_concat) else measures_concat
+        filtered_weights_concat = weights_concat[~np.isinf(measures_concat).any(axis=1)] if len(measures_concat) else weights_concat
+
+        n_clusters = self.quantiser.n_clusters
+        n_points = len(filtered_measures_concat)
+        if n_points < n_clusters:
+            # If not enough points to fit (including 0), let's arbitrarily put centers in [0, 1)^2
+            print(f"[Atol] had {n_points} points to fit {n_clusters} clusters, adding random points in [0, 1)^2.")
+            filtered_weights_concat = np.concatenate((filtered_weights_concat, np.ones(shape=(n_clusters - n_points))))
+            filtered_measures_concat = np.concatenate((filtered_measures_concat, np.random.random((n_clusters - n_points, 2))))
+
+        self.quantiser.fit(X=filtered_measures_concat, sample_weight=filtered_weights_concat)
```

Comment:
When the number of points equals the number of requested clusters, are clustering algorithms usually good at finding all of the points as centers? Another possibility, in the "too few points" case, would be to write to self.centers directly and skip the use of the quantiser.

Reply:
First of all thanks for all the great remarks.
I think on this specific question they are ok at that yes. And while I like your suggestion, if there's only say two points and we want five clusters, it _is_ probably highly valuable to take those two points as cluster centers and do whatever for the other three.
So I am still thinking but sort of converging into doing the following:
- filter more with np.unique (there is an overhead, because what we really only need to know is how many unique point up-to-`n_clusters` are there in the target set)
- fill in the rest with `[-np.inf, -np.inf]` (that would be closer to Landscapes) and for these centers write the inertias as 0 and hopefully that would not raise any errors at transform time.

Will work on that!

Reply:
> * filter more with np.unique (there is an overhead, because what we really only need to know is how many unique point up-to-`n_clusters` are there in the target set)

Yeah, I don't think there is a function that begins sorting or hash-tabling and stops once it has found k distinct elements... I think the common case will be that the first `n_clusters` points are distinct, maybe you could first check that with np.unique (does it actually work on points?), and only go for something slower in other cases?

Actually, do we really care? Having a duplicated feature is not much worse than having a feature that is always 0, and this is only for a degenerate case that really shouldn't happen anyway, so I don't think we should make too much effort beyond "warn and don't crash". I only mentioned the possibility of duplicate points in passing, but it does not bother me.

> * fill in the rest of the centers with `[-np.inf, -np.inf]` (that would be closer to Landscapes) and for these centers write the inertias as 0 and hopefully that would not raise any errors at transform time.

As you like. Unless someone else chimes in, I don't really care what points you use as filler.

Reply:
I agree that we do not really care, so I have dropped this!